File: src/java/ironfist/Referee.java
package ironfist;

import ironfist.generator.RecursiveDungeonGenerator;
import ironfist.math.Vector;
import ironfist.persistence.Persistence;
import ironfist.persistence.PersistenceException;

/**
 * DOCUMENT ME!
 * 
 * @author pmgremo
 */
public class Referee {

  private static final String CREATURE = "creature";
  private static final StairsPredicate stairsPredicate;
  private static final RecursiveDungeonGenerator generator;
  private static boolean run;

  static {
    generator = new RecursiveDungeonGenerator(System.currentTimeMillis());
    stairsPredicate = new StairsPredicate();
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   * 
   * @return DOCUMENT ME!
   */
  static public boolean run(Creature creature) {
    run = true;

    while (run) {
      Level level = creature.getLevel();
      level.run();

      if (!level.isActive()) {
        try {
          Persistence.put(level.getName(), level);
        } catch (Exception e) {
        }
      }
    }

    return run;
  }

  /**
   * DOCUMENT ME!
   */
  static public void stop() {
    run = false;
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   */
  public static void down(Creature creature) {
    Vector sourceLocation = creature.getCoordinate();
    Level sourceLevel = creature.getLevel();
    Tile tile = sourceLevel.get(sourceLocation);
    TileType type = tile.getTileType();

    if (type instanceof Floor) {
      Floor floor = (Floor) type;
      Portal portal = floor.getPortal();

      if (portal instanceof Stairs) {
        Stairs sourceStairs = (Stairs) portal;

        if (Stairs.DOWN.equals(sourceStairs.getDirection())) {
          Vector targetLocation = sourceStairs.getCoordinate();
          String targetName = sourceStairs.getLevelName();
          Tile targetTile = null;
          Level targetLevel = null;

          if (targetName == null) {
            String sourceName = sourceLevel.getName();
            targetName = Integer.toString(Integer.parseInt(sourceName) + 1);
            targetLevel = generator.generate(targetName);
            stairsPredicate.setDirection(Stairs.UP);
            targetTile = targetLevel.getRandom(stairsPredicate);

            Stairs targetStairs = (Stairs) ((Floor) targetTile.getTileType()).getPortal();
            targetStairs.setCoordinate(sourceLocation);
            targetStairs.setLevelName(sourceName);
            targetLocation = targetTile.getCoordinate();
            sourceStairs.setCoordinate(targetLocation);
            sourceStairs.setLevelName(targetName);
          } else {
            try {
              targetLevel = (Level) Persistence.get(targetName);
            } catch (Exception e) {
            }

            targetTile = targetLevel.get(targetLocation);
          }

          floor.setCreature(null);
          ((Floor) targetTile.getTileType()).setCreature(creature);
          creature.setCoordinate(targetLocation);
          creature.setLevel(targetLevel);
          targetLevel.setActive(true);
          sourceLevel.setActive(false);
        }
      }
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   */
  public static void up(Creature creature) {
    Vector sourceLocation = creature.getCoordinate();
    Level sourceLevel = creature.getLevel();
    Tile tile = sourceLevel.get(sourceLocation);
    TileType type = tile.getTileType();

    if (type instanceof Floor) {
      Floor floor = (Floor) type;
      Portal portal = floor.getPortal();

      if (portal instanceof Stairs) {
        Stairs sourceStairs = (Stairs) portal;

        if (Stairs.UP.equals(sourceStairs.getDirection())) {
          Vector targetLocation = sourceStairs.getCoordinate();
          String targetName = sourceStairs.getLevelName();
          Tile targetTile = null;
          Level targetLevel = null;

          if (targetName == null) {
            String sourceName = sourceLevel.getName();
            targetName = Integer.toString(Integer.parseInt(sourceName) + 1);
            targetLevel = generator.generate(targetName);
            stairsPredicate.setDirection(Stairs.DOWN);
            targetTile = targetLevel.getRandom(stairsPredicate);

            Stairs targetStairs = (Stairs) ((Floor) targetTile.getTileType()).getPortal();
            targetStairs.setCoordinate(sourceLocation);
            targetStairs.setLevelName(sourceName);
            targetLocation = targetTile.getCoordinate();
            sourceStairs.setCoordinate(targetLocation);
            sourceStairs.setLevelName(targetName);
          } else {
            try {
              targetLevel = (Level) Persistence.get(targetName);
            } catch (Exception e) {
            }

            targetTile = targetLevel.get(targetLocation);
          }

          floor.setCreature(null);
          ((Floor) targetTile.getTileType()).setCreature(creature);
          creature.setCoordinate(targetLocation);
          creature.setLevel(targetLevel);
          targetLevel.setActive(true);
          sourceLevel.setActive(false);
        }
      }
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   * @param direction
   *          DOCUMENT ME!
   */
  static public void move(Creature creature, Vector direction) {
    Vector oldCoordinate = creature.getCoordinate();
    Vector coordinate = oldCoordinate.add(direction);
    Level level = creature.getLevel();
    Tile tile = level.get(coordinate);
    TileType type = tile.getTileType();

    if (type instanceof Floor) {
      Floor floor = (Floor) type;
      Door door = floor.getDoor();

      if ((door == null) || door.isOpen()) {
        creature.setCoordinate(coordinate);
        floor.setCreature(creature);

        Tile oldTile = level.get(oldCoordinate);
        Floor oldFloor = (Floor) oldTile.getTileType();
        oldFloor.setCreature(null);
      }
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   * @param thing
   *          DOCUMENT ME!
   */
  static public void pickup(Creature creature, Thing thing) {
    Vector coordinate = creature.getCoordinate();
    Level level = creature.getLevel();
    Tile tile = level.get(coordinate);
    TileType type = tile.getTileType();

    if (type instanceof Floor) {
      Floor floor = (Floor) type;
      creature.addThing(thing);
      thing.setOwner(creature);
      floor.removeThing(thing);
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   * @param thing
   *          DOCUMENT ME!
   */
  static public void drop(Creature creature, Thing thing) {
    Vector coordinate = creature.getCoordinate();
    Level level = creature.getLevel();
    Tile tile = level.get(coordinate);
    TileType type = tile.getTileType();

    if (type instanceof Floor) {
      Floor floor = (Floor) type;
      creature.removeThing(thing);
      thing.setOwner(null);
      floor.addThing(thing);
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   * @param door
   *          DOCUMENT ME!
   */
  static public void open(Creature creature, Door door) {
    if ((door != null) && !door.isOpen()) {
      door.open();
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   * @param door
   *          DOCUMENT ME!
   */
  static public void close(Creature creature, Door door) {
    if ((door != null) && door.isOpen()) {
      door.close();
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   */
  static public void quit(Creature creature) {
    Persistence.delete(creature.getName());
    stop();
  }

  /**
   * DOCUMENT ME!
   * 
   * @param name
   *          DOCUMENT ME!
   * 
   * @return DOCUMENT ME!
   */
  static public Creature load(String name) {
    Creature result = null;

    try {
      Persistence.open(name);

      result = (Creature) Persistence.get(CREATURE);
    } catch (PersistenceException e) {
      e.printStackTrace();
    }

    return result;
  }

  /**
   * DOCUMENT ME!
   * 
   * @param creature
   *          DOCUMENT ME!
   */
  static public void save(Creature creature) {
    try {
      Persistence.put(CREATURE, creature);
    } catch (PersistenceException e) {
      e.printStackTrace();
    }

    stop();
  }

}

File: src/main/java/ukalus/Referee.kt
package ukalus

import org.apache.commons.math3.random.MersenneTwister
import org.apache.commons.math3.random.RandomAdaptor
import ukalus.level.dungeon.recursive.RecursiveDungeonGenerator
import ukalus.math.Vector2D
import ukalus.persistence.Persistence
import ukalus.persistence.PersistenceException

/**
 * DOCUMENT ME!

 * @author pmgremo
 */
object Referee {

    private val CREATURE = "creature"
    private val stairsPredicate: StairsPredicate = StairsPredicate()
    private val generator: RecursiveDungeonGenerator = RecursiveDungeonGenerator(RandomAdaptor(MersenneTwister(System.currentTimeMillis())))
    private var run: Boolean = false

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     * *
     * *
     * @return DOCUMENT ME!
     */
    fun run(creature: Creature): Boolean {
        run = true

        while (run) {
            val level = creature.level
            level!!.run()

            if (!level.isActive) {
                try {
                    Persistence.put(level.name, level)
                } catch (e: Exception) {
                }

            }
        }

        return run
    }

    /**
     * DOCUMENT ME!
     */
    fun stop() {
        run = false
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     */
    fun down(creature: Creature) {
        val sourceLocation = creature.coordinate
        val sourceLevel = creature.level
        val tile = sourceLevel!![sourceLocation!!]
        val type = tile!!.tileType

        if (type is Floor) {
            val floor = type
            val portal = floor.portal

            if (portal is Stairs) {
                val sourceStairs = portal as Stairs?

                if (Stairs.DOWN == sourceStairs!!.direction) {
                    var targetLocation: Vector2D = sourceStairs.coordinate!!
                    var targetName = sourceStairs.levelName
                    val targetTile: Tile?
                    var targetLevel: Level? = null

                    if (targetName == null) {
                        val sourceName = sourceLevel.name
                        targetName = Integer.toString(Integer.parseInt(sourceName) + 1)
                        targetLevel = generator.generate(targetName)
                        stairsPredicate.direction = Stairs.UP
                        targetTile = targetLevel!!.getRandom(stairsPredicate)

                        val targetStairs = (targetTile!!.tileType as Floor).portal as Stairs?
                        targetStairs!!.coordinate = sourceLocation
                        targetStairs.levelName = sourceName
                        targetLocation = targetTile.location
                        sourceStairs.coordinate = targetLocation
                        sourceStairs.levelName = targetName
                    } else {
                        try {
                            targetLevel = Persistence[targetName] as Level?
                        } catch (e: Exception) {
                        }

                        targetTile = targetLevel!![targetLocation]
                    }

                    floor.creature = null
                    (targetTile!!.tileType as Floor).creature = creature
                    creature.coordinate = targetLocation
                    creature.level = targetLevel
                    targetLevel.isActive = true
                    sourceLevel.isActive = false
                }
            }
        }
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     */
    fun up(creature: Creature) {
        val sourceLocation = creature.coordinate
        val sourceLevel = creature.level
        val tile = sourceLevel!![sourceLocation!!]
        val type = tile!!.tileType

        if (type is Floor) {
            val floor = type
            val portal = floor.portal

            if (portal is Stairs) {
                val sourceStairs = portal as Stairs?

                if (Stairs.UP == sourceStairs!!.direction) {
                    var targetLocation: Vector2D = sourceStairs.coordinate!!
                    var targetName = sourceStairs.levelName
                    val targetTile: Tile?
                    var targetLevel: Level? = null

                    if (targetName == null) {
                        val sourceName = sourceLevel.name
                        targetName = Integer.toString(Integer.parseInt(sourceName) + 1)
                        targetLevel = generator.generate(targetName)
                        stairsPredicate.direction = Stairs.DOWN
                        targetTile = targetLevel!!.getRandom(stairsPredicate)

                        val targetStairs = (targetTile!!.tileType as Floor).portal as Stairs?
                        targetStairs!!.coordinate = sourceLocation
                        targetStairs.levelName = sourceName
                        targetLocation = targetTile.location
                        sourceStairs.coordinate = targetLocation
                        sourceStairs.levelName = targetName
                    } else {
                        try {
                            targetLevel = Persistence[targetName] as Level?
                        } catch (e: Exception) {
                        }

                        targetTile = targetLevel!![targetLocation]
                    }

                    floor.creature = null
                    (targetTile!!.tileType as Floor).creature = creature
                    creature.coordinate = targetLocation
                    creature.level = targetLevel
                    targetLevel.isActive = true
                    sourceLevel.isActive = false
                }
            }
        }
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     * *
     * @param direction
     * *          DOCUMENT ME!
     */
    fun move(creature: Creature, direction: Vector2D) {
        val oldCoordinate = creature.coordinate
        val coordinate = oldCoordinate!!.plus(direction)
        val level = creature.level
        val tile = level!![coordinate]
        val type = tile!!.tileType

        if (type is Floor) {
            val floor = type
            val door = floor.door

            if (door == null || door.isOpen) {
                creature.coordinate = coordinate
                floor.creature = creature

                val oldTile = level[oldCoordinate]
                val oldFloor = oldTile!!.tileType as Floor
                oldFloor.creature = null
            }
        }
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     * *
     * @param thing
     * *          DOCUMENT ME!
     */
    fun pickup(creature: Creature, thing: Thing) {
        val coordinate = creature.coordinate
        val level = creature.level
        val tile = level!![coordinate!!]
        val type = tile!!.tileType

        if (type is Floor) {
            creature.addThing(thing)
            thing.owner = creature
            type.removeThing(thing)
        }
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     * *
     * @param thing
     * *          DOCUMENT ME!
     */
    fun drop(creature: Creature, thing: Thing) {
        val coordinate = creature.coordinate
        val level = creature.level
        val tile = level!![coordinate!!]
        val type = tile!!.tileType

        if (type is Floor) {
            creature.removeThing(thing)
            thing.owner = null
            type.addThing(thing)
        }
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     * *
     * @param door
     * *          DOCUMENT ME!
     */
    fun open(creature: Creature, door: Door?) {
        if (door != null && !door.isOpen) {
            door.open()
        }
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     * *
     * @param door
     * *          DOCUMENT ME!
     */
    fun close(creature: Creature, door: Door?) {
        if (door != null && door.isOpen) {
            door.close()
        }
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     */
    fun quit(creature: Creature) {
        Persistence.delete(creature.name!!)
        stop()
    }

    /**
     * DOCUMENT ME!

     * @param name
     * *          DOCUMENT ME!
     * *
     * *
     * @return DOCUMENT ME!
     */
    fun load(name: String): Creature {
        Persistence.open(name)
        return Persistence[CREATURE] as Creature
    }

    /**
     * DOCUMENT ME!

     * @param creature
     * *          DOCUMENT ME!
     */
    fun save(creature: Creature) {
        try {
            Persistence.put(CREATURE, creature)
        } catch (e: PersistenceException) {
            e.printStackTrace()
        }

        stop()
    }

}